feat(medusa): Stock location module

carlos-r-l-rodrigues opened the pull request:
What:
Module to manage stock location

Fixes CORE-916

carlos-r-l-rodrigues commented on scripts/build-openapi.js:
```diff
@@ -3,111 +3,141 @@
-  ["./packages/medusa/src/models", "./packages/medusa/src/api/middlewares" , "./packages/medusa/src/api/routes/store"],
+  [
+    "./packages/medusa/src/models",
+    "./packages/medusa/src/types",
```
Note:
Adding `./packages/medusa/src/types` to `swaggerInline` was the only change in this file.

carlos-r-l-rodrigues commented on packages/medusa/src/migrations/1666251508718-stock_location.ts:
```diff
@@ -0,0 +1,23 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
```
todo: remove this migration once https://github.com/medusajs/medusa/pull/2883 is merged

pKorsholm commented on packages/medusa/src/api/routes/admin/stock-locations/update-stock-location.ts:
```diff
@@ -0,0 +1,183 @@
+ *     example: {car: "white"}
+ *   address:
+ *     type: object
+ *     properties:
+ *       address_1:
+ *         type: string
+ *         description: Stock location address
+ *         example: 35, Jhon Doe Ave
+ *       address_2:
+ *         type: string
+ *         description: Stock location address' complement
+ *         example: apartment 4432
+ *       city:
+ *         type: string
+ *         description: Stock location address' city
+ *         example: Mexico city
+ *       country_code:
+ *         type: string
+ *         description: Stock location address' country
+ *         example: MX
+ *       phone:
+ *         type: string
+ *         description: Stock location address' phone number
+ *         example: +1 555 61646
+ *       postal_code:
+ *         type: string
+ *         description: Stock location address' postal code
+ *         example: HD3-1G8
+ *       province:
+ *         type: string
+ *         description: Stock location address' province
+ *         example: Sinaloa
+ *       metadata:
+ *         type: object
+ *         description: An optional key-value map with additional details
+ *         example: {car: "white"}
```
suggestion(non-blocking): extract as separate OAS object

carlos-r-l-rodrigues replied:
referencing StockLocationAddressInput would be ok?

pKorsholm commented on packages/medusa/src/loaders/module-definitions/definitions.ts:
```diff
@@ -4,7 +4,7 @@ export const MODULE_DEFINITIONS: ModuleDefinition[] = [
     key: "stockLocationService",
     registrationName: "stockLocationService",
-    defaultPackage: false,
+    defaultPackage: "@medusajs/stock-location",
```
I think this should still be `false` as long as we keep developing before release since we have opted for not having a dedicated feature flag. @srindom wdyt?

pKorsholm commented on packages/stock-location/jest.config.js:
```diff
@@ -0,0 +1,23 @@
+module.exports = {
+  //moduleNameMapper: {
+  //  "^highlight.js$": `<rootDir>/node_modules/highlight.js/lib/index.js`,
+  //},
+  //snapshotSerializers: [`jest-serializer-path`],
+  // collectCoverageFrom: coverageDirs,
+  //reporters: process.env.CI
+  //  ? [[`jest-silent-reporter`, { useDots: true }]].concat(
+  //      useCoverage ? `jest-junit` : []
+  //    )
+  //  : [`default`].concat(useCoverage ? `jest-junit` : []),
```
rm

pKorsholm commented on packages/stock-location/src/services/stock-location.ts:
```diff
@@ -0,0 +1,216 @@
+import { getConnection, EntityManager } from "typeorm"
```
suggestion: add jsdocs

srindom replied:
this is todo :)

pKorsholm reviewed (commented):
Very nice! I only have a couple of suggestions but nothing blocking 😄

carlos-r-l-rodrigues commented on packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts:
```diff
@@ -0,0 +1,178 @@
+ *     type: object
+ *     description: An optional key-value map with additional details
+ *     example: {car: "white"}
+ *   address:
+ *     type: object
+ *     properties:
+ *       address_1:
```
I thought about it, but `class StockLocationAddress` isn't exported so I was unsure what to do.
Also thought about reference it to `StockLocationAddressInput` defined on `/types`. That could be ok, considering the input will have to respect that definition.

carlos-r-l-rodrigues commented on packages/medusa/src/api/routes/admin/stock-locations/list-stock-locations.ts:
```diff
@@ -0,0 +1,179 @@
+ *     description: Date comparison for when resulting collections were updated.
+ *     schema:
+ *       type: object
+ *       properties:
+ *         lt:
+ *            type: string
+ *            description: filter by dates less than this date
+ *            format: date
+ *         gt:
+ *            type: string
+ *            description: filter by dates greater than this date
+ *            format: date
+ *         lte:
+ *            type: string
+ *            description: filter by dates less than or equal to this date
+ *            format: date
+ *         gte:
+ *            type: string
+ *            description: filter by dates greater than or equal to this date
+ *            format: date
```
good call. That is replicated in multiple places

srindom commented on packages/medusa/src/services/sales-channel-location.ts:
```diff
@@ -0,0 +1,80 @@
+import { EntityManager } from "typeorm"
```
todo (this-file): add JSDoc

srindom reviewed (commented):
A few small todos :)

srindom reviewed (approved):
LGTM!

srindom commented:
Feel free to merge - just want to remind you about the your comment in the migration file. Do you need to delete the migration before merging?

carlos-r-l-rodrigues commented:
> Feel free to merge - just want to remind you about the your comment in the migration file. Do you need to delete the migration before merging?

I removed it. I fixed a few conflicts but I will ask you @srindom and @pKorsholm to take a look at the  changes on `claim.ts` service that I did.

srindom commented on packages/medusa/src/services/claim.ts:
```diff
@@ -332,9 +332,30 @@ export default class ClaimService extends TransactionBaseService {
         let newItems: LineItem[] = []
+
         if (isDefined(additional_items)) {
+          await Promise.all(
```
todo: I would remove this check. It is generally business users who create claims and they prefer having the control of forcing a claim through despite there not being stock at this time. E.g., if they know that the items will be restocked soon. 

I can see other use cases where merchants might not want this behavior, but would decouple the check from the claim creation and instead look toward a customization option in a future operation.

srindom replied:
One thing that may make sense to add now is reserving stock for claims. (Would still decouple this from the claim creation and add it in the API handler instead.)

carlos-r-l-rodrigues replied:
that is being handled in the current flow that Phil covered. I believe it is better to have another PR to change the place where the stock is being allocated.
By the way, are we keeping `reserve` or should we rename everything to `allocate`

srindom reviewed (commented):
@carlos-r-l-rodrigues I have added notes on claim.ts

srindom commented on packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts:
```diff
@@ -0,0 +1,145 @@
+    "stockLocationService"
+  )
+
+  const stockLocation = await locationService.create(
+    req.validatedBody as AdminPostStockLocationsReq
+  )
```
This would be an excellent use case for middleware :) 

```
const locationService = req.scope.resolve("stockLocationService")

if (!locationService) {
  // throw: Stock Locations are not enabled. Please add a Stock Location module to enable this functionality.
}

next()
```

pKorsholm replied:
`moduleRegisteredMiddleware`?

adrien2p commented on packages/stock-location/src/services/stock-location.ts:
```diff
@@ -0,0 +1,266 @@
+   * @param {string} id - The ID of the stock location to delete.
+   * @returns {Promise<void>} - An empty promise.
+   */
+  async delete(id: string): Promise<void> {
```
todo: Missing context for manager

adrien2p commented on packages/medusa/src/services/sales-channel-location.ts:
```diff
@@ -51,19 +56,20 @@ class SalesChannelLocationService extends TransactionBaseService {
       .withTransaction(manager)
       .retrieve(salesChannelId)
-    const stockLocation = await this.stockLocationService_.retrieve(locationId)
+    const stockLocation = await this.stockLocationService.retrieve(locationId)
```
todo: Depending on the end comment about the missing context for the transaction, we should pass the manager as part of the context to avoid inconsistency with the actual approach. wdyt?

adrien2p replied:
same for the update

adrien2p reviewed (changes requested):
Great works 🎉  I still have a few things though

fPolic commented on packages/medusa-js/src/resources/admin/stock-locations.ts:
```diff
@@ -0,0 +1,78 @@
+    customHeaders: Record<string, any> = {}
+  ): ResponsePromise<AdminStockLocationsRes> {
+    const path = `/admin/stock-locations`
+    return this.client.request("GET", path, payload, {}, customHeaders)
```
```suggestion
    return this.client.request("POST", path, payload, {}, customHeaders)
```

todo: comments need to be updated

fPolic reviewed (commented):
Great work!
Just a small change in the client is needed but LGTM!

carlos-r-l-rodrigues commented on packages/medusa/src/interfaces/services/event-bus.ts:
```diff
@@ -0,0 +1,3 @@
+export interface IEventBusService {
+  emit(event: string, data: any): Promise<void>
```
the class StockLocationService is using it

carlos-r-l-rodrigues commented on packages/medusa/src/types/stock-location.ts:
```diff
@@ -1,21 +1,121 @@
+  postal_code?: string | null
+  province?: string | null
+  metadata?: Record<string, unknown> | null
+  created_at: string | Date
```
These types are related to the module, if any external module wants to communicate with that, this is the type that has to be used. Maybe we can find a different name in the future to identify these types, but for now I will keep it like this.

carlos-r-l-rodrigues commented on packages/stock-location/src/services/stock-location.ts:
```diff
@@ -0,0 +1,266 @@
+    const locationRepo = manager.getRepository(StockLocation)
+
+    const query = buildQuery(selector, config)
+    return await locationRepo.find(query)
```
Count will add an extra DB query to count the records before returning them.

adrien2p commented on packages/stock-location/src/migrations/schema-migrations/1665749860179-setup.ts:
```diff
@@ -0,0 +1,94 @@
+        "phone" TEXT,
+        "province" TEXT,
+        "postal_code" TEXT,
+        "metadata" JSONB,
```
I was thinking of '{}'::JSONB

carlos-r-l-rodrigues commented on packages/stock-location/src/services/stock-location.ts:
```diff
@@ -0,0 +1,266 @@
+
+      const result = await locationRepo.save(loc)
+
+      await this.eventBusService_.emit(StockLocationService.Events.CREATED, {
```
This will need to be properly handled by the Event Outbox pattern that @olivermrbl is planning.

adrien2p replied:
Can we add a todo? My concern is that since this feature is not guarded, anybody can install the module and end up with discrepancies. Do we have a warning somewhere ? I might have missed it

carlos-r-l-rodrigues commented on packages/stock-location/src/services/stock-location.ts:
```diff
@@ -0,0 +1,266 @@
+
+  async update(
+    itemId: string,
+    updateData: UpdateStockLocationInput
```
Which consumer you mean? As this will be an independent module, a transaction orchestrator will be needed to revert failed transaction involving this module and the core.

carlos-r-l-rodrigues replied:
only on the JS client, there is a * @experimental on the JSDoc

adrien2p commented on packages/medusa/src/api/routes/admin/stock-locations/index.ts:
```diff
@@ -16,12 +16,20 @@ import {
 export default (app) => {
   app.use("/stock-locations", route)
 
+  app.use(
```
suggestion(non-blocking): This can be moved directly in the top app.use

adrien2p commented on packages/medusa/src/api/middlewares/check-registered-modules.ts:
```diff
@@ -0,0 +1,15 @@
+  [serviceName: string]: string
+}): (req: Request, res: Response, next: NextFunction) => Promise<void> {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    for (const service of Object.keys(services)) {
```
todo: should we add a test, I think that in order for the resolve to not throw, we need to add the option `allowUnregistered` otherwise it will not return undefined but throw instead. Can I get you to try please

adrien2p reviewed (approved):
LGTM 🎉 Just one thing to look at before merging please
